### PR TITLE
stats: fix scope cache consistency issue when scope memory is recyled

### PR DESCRIPTION
**Pull request by mattklein123**
The memory allocator can reuse a memory address immediately, leading to
a situation in which TLS references are used for an already destroyed
scope. Instead, an incrementing ID is used as the cache key.

Fixes https://github.com/envoyproxy/envoy/issues/3214

*Risk Level*: Low 
*Testing*: Coming up with a test case for this would involve mocking the memory
allocator. I'm not sure it's really worth it. IMO the existing tests are good enough
along with code inspection and manual verification that this fixes the reported
issue.
*Docs Changes*: N/A
*Release Notes*: N/A

### CODE REVIEW

**Comment by mattklein123** [comment]
cc @julia-stripe @jmillikin-stripe please verify the fix if you can.

**Comment by julia-stripe** [comment]
amazing, thank you so much for coming up with a patch so quickly! I tested and this appears to fix the segfault.

File: source/common/stats/thread_local_store.h [hunk 1]
```diff
@@ -95,13 +96,22 @@ class ThreadLocalStoreImpl : public StoreRoot {
...
+    const uint64_t scope_id_;
     ThreadLocalStoreImpl& parent_;
     const std::string prefix_;
     TlsCacheEntry central_cache_;
   };
 
   struct TlsCache : public ThreadLocal::ThreadLocalObject {
-    std::unordered_map<ScopeImpl*, TlsCacheEntry> scope_cache_;
+    // The TLS scope cache is keyed by scope ID. This is used to avoid complex circular references
+    // during scope destruction. An ID is required vs. using the address of the scope pointer
```
[the remark this answers is not in the document]
Comment (mattklein123) [reply source/common/stats/thread_local_store.h]: Scopes can be deleted from any thread. Clearing the cache in the TLS slot on that thread would be easy, but might not still solve the problem (i.e., if you deleted on a worker but then created a scope on the main thread). To remain lockless, this is why posting is used which exposes this issue. Switching to a unique incrementing ID makes sure that there can never be this type of cache re-reference.

File: source/common/stats/thread_local_store.h [hunk 2]
```diff
@@ -95,13 +96,22 @@ class ThreadLocalStoreImpl : public StoreRoot {
...
+    // because it's possible that the memory allocater will recyle the scope pointer immediately
+    // upon destruction, leading to a situation in which a new scope with the same address is used
+    // to reference the cache, and then subsequently cache flushed, leaving nothing in the central
+    // store.
+    std::unordered_map<uint64_t, TlsCacheEntry> scope_cache_;
```
[the remark this answers is not in the document]
Comment (mattklein123) [reply source/common/stats/thread_local_store.h]: Scopes are global. This is TLS.

File: source/common/stats/thread_local_store.h (hunk 1 above)
[the remark this answers is not in the document]
Comment (jmarantz) [reply source/common/stats/thread_local_store.h]: Got it, sounds good to me.  Can you add a note about remaining lockless to the comment?

**Comment by julia-stripe** [comment]
Should this comment also be updated?  https://github.com/envoyproxy/envoy/blob/master/source/common/stats/thread_local_store.h#L37-L40 

One thing that's still not clear to me is -- the comment about the race says "if [the race] happens the cache will be repopulated on the next access.", which seems not to have happened in this case, and I haven't figured out why not yet.

**Comment by mattklein123** [comment]
@jmarantz @mrice32 @julia-stripe added/fixed comments. I had forgotten about that other comment that is now outdated. I guess I should listen to myself when I write comments and read them later. ;)

File: source/common/stats/thread_local_store.h [hunk 3]
```diff
@@ -23,6 +23,7 @@ namespace Stats {
  * - Overallaping scopes with proper reference counting (2 scopes with the same name will point to
```
Comment (jmarantz) [inline source/common/stats/thread_local_store.h]: Overlapping

while you are in here :)

**Comment by julia-stripe** [comment]
One more question (for my own edification) -- why is it possible for there to be dangling references to a deleted `Counter` object (or other stats object)? I see that there's a `CounterSharedPtr` type being used in `thread_local_store.cc`, but it looks like classes that refer to these counters (eg `ClusterInfoImpl`) use a reference instead of a `shared_ptr`. Why is that?

**Comment by julia-stripe** [comment]
> as users could hold a stat reference beyond the life of a scope which is technically wrong

is the idea if a user has a stat reference beyond the life of a scope, then the best thing to do is to segfault?

**Comment by mattklein123** [comment]
> is the idea if a user has a stat reference beyond the life of a scope, then the best thing to do is to segfault?

Although it's never nice to crash, sometimes this is the reality of bugs and a clear indication that something needs to be fixed.

**Pull request merged**